### PR TITLE
SIMD: Merge Loongson-MMI into MIPS

### PR DESCRIPTION
**Pull request by FlyGoat**
So we can apply loongson mmi to generic MIPS builds and detect them at runtime.
That also allow us share MIPS MSA SIMD implemention later.

### CODE REVIEW

**Comment by dcommander** [comment]
In order to integrate this, I need temporary access to a Loongson and a non-Loongson MIPS machine for testing.

**Comment by dcommander** [comment]
Please follow up through e-mail in order to obtain my SSH public key:
https://libjpeg-turbo.org/About/Contact

**Comment by FlyGoat** [comment]
email sent.
I don't have a MIPS32 machine with DSPR2 support. My Loongson-3A3000 supports both DSPr2 and MMI.
But current DSPr2 code only supports 32bit system. For 64bit systems it needs some modifications like 64bit pointers.

File: simd/CMakeLists.txt [hunk 1]
```diff
@@ -30,9 +30,6 @@ if(CPU_TYPE STREQUAL "x86_64")
   if(CYGWIN)
     set(CMAKE_ASM_NASM_OBJECT_FORMAT win64)
   endif()
-  if(CMAKE_C_COMPILER_ABI MATCHES "ELF X32")
-    set(CMAKE_ASM_NASM_OBJECT_FORMAT elfx32)
-  endif()
```
Comment (dcommander) [inline simd/CMakeLists.txt]: Why did you remove this?  It has nothing to do with MIPS.  You just broke the build on a completely unrelated platform.

**Comment by dcommander** [comment]
You seem to be trying to build both DSPr2 and MMI code into the same libjpeg-turbo binary, but the build system modifications you made do not achieve that end.  `CPU_TYPE` can only ever have one value, and the DSPr2 code will only be included if `CPU_TYPE` is `mips` or `mipsel`.  However, the MMI code will never be included if `CPU_TYPE` is `mips` or `mipsel`.  Thus, there is no value of `CPU_TYPE` for which both DSPr2 and MMI will be included, and thus this PR does not achieve anything that could not already be achieved.

File: simd/CMakeLists.txt (hunk 1 above)
Comment (FlyGoat) [reply simd/CMakeLists.txt]: Sorry. That was a copy and paste issue.

**Comment by FlyGoat** [comment]
That was intentional. As current MMI code only runs on 64-bit platform while DSPr2 only runs on 32-bit platform.
This PR is achieving: Generic MIPS64 build on distros like Debian can benefit from MMI optimization on Loongson machines. And I see another person is submitting MIPS MSA support. Newer Loongson processor support that SIMD too. We can share that implementation.

**Comment by dcommander** [comment]
I was under the impression that some MIPS64 platforms have support for DSP r2.

**Comment by FlyGoat** [comment]
Yes, some MIPS64 platforms do support DSPr2, but our current codebase doesn't fit 64-bit platform.

**Comment by FlyGoat** [comment]
What will be the next?

**Comment by dcommander** [comment]
What do you mean by “our current code base”?

**Comment by dcommander** [comment]
In libjpeg-turbo, 32-bit and 64-bit SIMD extensions should never share the same directory. So if our existing DSP r2 SIMD extensions cannot work with 64-bit, and if our existing MMI extensions cannot work with 32-bit, then I would suggest moving the MMI code into a directory called “mips64”. This could be done in conjunction with integrating the submitted MSA extensions.

**Comment by FlyGoat** [comment]
jsimd_dspr2.S can't handle 64-bit pointers, that's the only issue.
Should I submit another PR that only adding runtime detect for MMI? I just noticed that all of the generic MIPS64 Linux distros can't be benefited from MMI optimization. That's why I propose this PR.

Thanks for your help~

**Comment by dcommander** [comment]
That makes sense. Submit another PR with run-time detection and the files moved from loongson/ to mips64/, and I will integrate that along with the MSA optimizations (the latter is waiting on funding.)
